pep-0522.txt

```diff
@@ -0,0 +1,239 @@
+random numbers), PEP 506 collected several of the one line recipes based
+on the lower level ``os.urandom()`` API into a new ``secrets`` module.
+
+However, this guidance has also come with a longstanding caveat: developers
+writing security sensitive software at least for Linux, and potentially for
+some other \*BSD systems, may need to wait until the operating system's
+random number generator is ready before relying on it for security sensitive
+operations.
+
+Unfortunately, there's currently no clear indicator to developers that their
+software may not be working as expected when run early in the Linux boot
+process, or on hardware without good sources of entropy to seed the operating
+system's random number generator: due to the behaviour of the underlying
+``/dev/urandom`` device, ``os.urandom()`` on Linux returns a result either way,
+and it takes extensive statistical analysis to show that a security
+vulnerability exists.
+
+By contrast, if ``BlockingIOError`` is raised in those situations, then
+developers can easily choose their desired behaviour:
+
+1. Loop until the call succeeds (security sensitive)
+2. Switch to using the random module (non-security sensitive)
+3. Switch to reading ``/dev/urandom`` directly (non-security sensitive)
+
+
+Why now?
+--------
+
+The main reason is because the 3.5 SipHash initialisation bug causing a deadlock
+when attempting to run Python scripts during the Linux init process resulted in
+a rash of proposals to add *new* APIs like ``getrandom()``, ``urandom_block()``,
+``pseudorandom()`` and ``cryptorandom()`` to the ``os`` module and to start
+trying to educate users on when they should call those APIs instead of
+``os.urandom()``.
+
+This is a *really* obscure problem, and we definitely shouldn't clutter up the
+standard library with new APIs without a compelling reason, especially with the
+``secrets`` module already being added as the "use this and don't worry about
+the low level details" for developers that don't need to worry about versions
+prior to Python 3.6.
+
+However, it's also the case that low cost ARM devices are becoming increasingly
+prevalent, with a lot of them running Linux, and a lot of folks writing
+Python applications that run on those devices. That creates an opportunity to
+take an obscure security problem that requires a lot of knowledge about
+Linux boot processes and secure random number generation and turn it into a
+relatively mundane and easy-to-find-in-an-internet-search runtime exception.
+
+
+Background
+==========
+
+On operating systems other than Linux, ``os.urandom()`` may already block
+waiting for the operating system's random number generator to be ready.
+
+On Linux, even when the operating system's random number generator doesn't
+consider itself ready for use in security sensitive operations, it will return
+random values based on the entropy it as available.
+
+This behaviour is potentially problematic, so Linux 3.17 added a new
```

Comment:
Notably the new interface to the urandom pool did *not* include any way to get the old "get predictable numbers instead of block" functionality. It is not possible to insecurely use the API, you either:

* Use urandom, blocking until the pool is initialized (default)
* Use random, blocking whenever the pool thinks it's entropy is low (``GRND_RANDOM``).
* Use urandom, returning a ``-1`` and setting eerno to ``EAGAIN`` if the pool isn't initialized (``GRND_NONBLOCK``).